paasta_tools/kubernetes_tools.py

```diff
@@ -1516,8 +1520,69 @@ def get_pod_volumes(
         if crypto_volume:
             pod_volumes.append(crypto_volume)
 
+        datastore_credentials_secrets_volume = (
+            self.get_datastore_credentials_secrets_volume()
+        )
+        if datastore_credentials_secrets_volume:
+            pod_volumes.append(datastore_credentials_secrets_volume)
+
         return pod_volumes
 
+    def get_datastore_credentials(self) -> DatastoreCredentialsConfig:
+        datastore_credentials = self.config_dict.get("datastore_credentials", {})
+        return datastore_credentials
+
+    def get_datastore_credentials_secret_name(self) -> str:
+        return _get_secret_name(
+            self.get_namespace(),
+            "datastore-credentials",
+            self.get_service(),
+            self.get_instance(),
+        )
+
+    def get_datastore_credentials_volume_name(self) -> str:
+        """
+        Volume names must abide to DNS mappings of 63 chars or less, so we limit it here and replace _ with --.
+        """
+        prefix = "datastore-credentials-volume"
+        service_name = self.get_sanitised_service_name()
+        instance_name = self.get_sanitised_instance_name()
+        return limit_size_with_hash(
+            name=f"{prefix}-{service_name}-{instance_name}".replace("_", "--")
+        )
+
+    def get_datastore_credentials_secrets_volume(self) -> List[V1Volume]:
+        """
+        All credentials are stored in 1 Kubernetes Secret, which are mapped on an item->path
+        structure to /datastore/<datastore>/<credential>/<password file>.
+        """
+        datastore_credentials = self.get_datastore_credentials()
```

Comment:
just as below, I think we'll want to do what we do for boto/crypto secrets and ensure that the secret exists (by checking for the existence of the signature)